apps/axistream/src/message.hh

```diff
@@ -0,0 +1,498 @@
+namespace message {
+
+enum class Sign {
+    UNSIGNED, SIGNED
+};
+
+enum class Endianness {
+    BIG, LITTLE
+};
+
+template <typename T>
+T flip_endianness(const T& x) {
+    static_assert(T::width % util::BITS_PER_BYTE == 0, "Type must be a integer number of bytes");
+    HLS_PRAGMA(HLS inline);
+    const int num_bytes = T::width / util::BITS_PER_BYTE;
+    T out = 0;
+    for (int i = 0; i < num_bytes; i++) {
+        HLS_PRAGMA(HLS unroll);
+        int j = num_bytes - i - 1;
+        out(util::BITS_PER_BYTE * (i + 1) - 1, util::BITS_PER_BYTE * i) = x(util::BITS_PER_BYTE * (j + 1) - 1, util::BITS_PER_BYTE * j);
+    }
+    return out;
+}
+
+/*
+ * Endianness Converter
+ */
+template <Endianness E>
+struct EndiannessConverter {};
+
+template <>
+struct EndiannessConverter<Endianness::LITTLE> {
+    template <int W>
+    static ap_uint<W> convert(const ap_uint<W>& x) {
+        HLS_PRAGMA(HLS inline);
+        return x;
+    }
+
+};
+
+template <>
+struct EndiannessConverter<Endianness::BIG> {
+    template <int W>
+    static ap_uint<W> convert(const ap_uint<W>& x) {
+        HLS_PRAGMA(HLS inline);
+        return flip_endianness(x);
+    }
+};
+
+/*
+ * Helper type to select between ap_uint and ap_int
+ */
+template <int W, Sign S>
+struct TypeSelector {};
+
+template <int W>
+struct TypeSelector<W, Sign::SIGNED> {
+    typedef ap_int<W> type;
+};
+
+template <int W>
+struct TypeSelector<W, Sign::UNSIGNED> {
+    typedef ap_uint<W> type;
+};
+
+template <> struct TypeSelector<8, Sign::SIGNED> { typedef int8_t type; };
+template <> struct TypeSelector<16, Sign::SIGNED> { typedef int16_t type; };
+template <> struct TypeSelector<32, Sign::SIGNED> { typedef int32_t type; };
+template <> struct TypeSelector<64, Sign::SIGNED> { typedef int64_t type; };
+template <> struct TypeSelector<8, Sign::UNSIGNED> { typedef uint8_t type; };
+template <> struct TypeSelector<16, Sign::UNSIGNED> { typedef uint16_t type; };
+template <> struct TypeSelector<32, Sign::UNSIGNED> { typedef uint32_t type; };
+template <> struct TypeSelector<64, Sign::UNSIGNED> { typedef uint64_t type; };
+
+/*
+ * Type definitions. All types must define:
+ * width: the width in bits
+ * raw: the type of a raw field - ap_uint<width>. Note that even
+ * though the raw value of a field might have the same type as the nominal
+ * type (given by type, below), it may not be equal in value.
+ * For example, the endianness may be flipped.
+ * type: the type of a field when used normally. For integer fields,
+ * this is ap_int<width> or ap_uint<width>. For enum fields,
+ * this is the enum type. For composite fields (fields with a struct
+ * or union type), this is a message of that type.
+ * to_raw: function from the nominal type to the raw type
+ * from_raw: function from the raw type to the nominal type
+ */
+
+template <int W, Sign S, Endianness E>
+struct Int {
+    static const int width = W;
+
+    typedef ap_uint<width> raw;
+    typedef typename TypeSelector<W, S>::type type;
```

Comment:
OK, so one advantage of this is that you have the ability to define arbitrary width fields.  Somewhat purposely, I had limited things in the original library to being 8-bit granularity.  One reason for this is that muxing gets expensive, and I wanted to ensure that the muxing wasn't hugely expensive for very wide data paths, even if constant offsets could not be proven.

Reply:
Within a message, every field has a constant offset - the offset is computed at compile time via the templates. So reading a field from a message should be a nop - a field is a named slice.